docs/design/0012-generic-association.md

```diff
@@ -0,0 +1,275 @@
+
+ * `username`: username of the user as it exists in Elasticsearch
+ * `password`: clear text version of the password, the hashed version is maintained, reconciled and is made available by the generic association controller to the Elasticsearch controller.
+ * `elasticsearch-host`: hostname of the Elasticsearch cluster
+ * `elasticsearch-port`: port of the Elasticsearch cluster
+ * `elasticsearch-ca`: Certificate of the Elasticsearch CA
+ * `kibana-host`: hostname to be used to access Kibana
+ * `kibana-port`: port of Kibana
+ * `kibana-ca`: certificate of the Kibana CA
+ * `apmserver-url`: url to be used to access the APM Server
+ * `apmserver-token`: token to be used to access the APM Server
+ * `apmserver-ca`: certificate of the APM Server CA
+
+## Injecting association data automatically
+
+#### Option #1: Updating the Pod's parent resource
+
+Two different approaches might be used in order to populate the environment variables and bind the `Secret` automatically in a `Pod`.
+
+In the first one ECK would insert some environment variables in the `Pod`'s parent resource:
+
+Here is the `Deployment` as it would be created by a user:
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: my-application
+spec:
+  selector:
+    matchLabels:
+      app: test
+  template:
+    metadata:
+      labels:
+        app: test
+    spec:
+      containers:
+      - name: container1
+        image: ...
+      - name: container2
+        image: ...
+```
+
+Here is the `Deployment` as it would be updated by ECK:
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: my-application
+spec:
+  selector:
+    matchLabels:
+      app: test
+  template:
+    metadata:
+      annotations:
+        elastic.config.checksum: 585696 # allows an automatic rotation
+      labels:
+        app: test
+    spec:
+      containers:
+      - name: container1
+        image: ...
+        env:
+          - name: ELASTICSEARCH_USERNAME
+            valueFrom:
+              secretKeyRef:
+                name: <automatic-generated-name>
+                key: username
+        volumeMounts:
+        - name: elastic-volume
+          mountPath: /mnt/elastic
+      - name: container2
+        image: ...
+        volumeMounts:
+        - name: elastic-volume
+          mountPath: /mnt/elastic
+      volumes:
+      - name: elastic-volume
+        secret:
+          secretName: <automatic-generated-name>
+```
+
+  * Pros:
+    * The configuration checksum can be injected in the pod template to allow an automatic restart of the `Pods` 
+    * ECK can inject and manage some "well-known" environment variables, some good examples are the ones used in the context of the APM Server (`ELASTIC_APM_SERVER_URLS`, `ELASTIC_APM_SECRET_TOKEN`...) 
+  * Cons:
+    * It could be considered as "intrusive"
```

Comment:
++ I don't like much the idea of ECK updating the spec of other Pods/Deployments/StatefulSets outside the scope of a MutatingAdmissionWebhook.

Reply:
I have the same feeling, it sounds odd. But I'm not sure I have a technical argument against it 🤷‍♂️